File: LinkedList/reverse-linked-list-2/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        # edge cases, list is blank, list has 1 item or left === right just return the list
        if not head or not head.next or left == right:
            return head
        # step 1, locate left and the previous to left, we will call these node1 and prev1 
        node1 = head
        prev1 = None
        for i in range(1, left):
            prev1 = node1
            node1 = node1.next
        # step 2, locate right and the next pointer of right, we will call these node2 and next2
        node2 = head
        next2 = None
        for i in range(1, right):
            node2 = node2.next
        next2 = node2.next
        
        # step 3, sever the ties! so prev1.next should be None and node2.next should be none if they exist
        if prev1:
            prev1.next = None
        if node2:
            node2.next = None
        
        # reverse the middle section, the middle section head is node1
        curr = node1
        prev = None
        nextr = None
        while curr:
            nextr = curr.next
            curr.next = prev
            prev = curr
            curr = nextr
        
        # alrighty so now prev is the head of the middle section and node1 = tail, time to attach
        if prev1:
            prev1.next = prev
        if node1:
            node1.next = next2
        return head if prev1 else prev
        
        
        

File: LinkedList/reverse-linked-list-2/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function (head, left, right) {
  // edge cases: list is empy, list of one, left = right
  if (!head || !head.next || left === right) {
    return head;
  }
  // step one, get node at position left and previous node
  let node1 = head;
  let prev1 = null;
  for (let i = 1; i < left; i++) {
    prev1 = node1;
    node1 = node1.next;
  }

  // step 2, get node at position right and next node
  let node2 = head;
  let next2 = null;
  for (let i = 1; i < right; i++) {
    node2 = node2.next;
  }
  next2 = node2.next;

  // step 3, make the cuts
  if (prev1) {
    prev1.next = null;
  }
  if (node2) {
    node2.next = null;
  }
  // reverse mid section
  let curr = node1,
    next = null,
    prev = null;
  while (curr) {
    next = curr.next;
    curr.next = prev;
    prev = curr;
    curr = next;
  }
  // at this juncture, prev is the reversed list and node1 is the tail so we need to attach
  if (prev1) {
    prev1.next = prev;
  }
  node1.next = next2;

  return prev1 ? head : prev;
};
